pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py

```diff
@@ -193,50 +193,50 @@ def _commit(self):
                 self._status = base.BatchStatus.SUCCESS
                 return
 
-            # Begin the request to publish these messages.
-            # Log how long the underlying request takes.
-            start = time.time()
-
-            try:
-                response = self._client.api.publish(self._topic, self._messages)
-            except google.api_core.exceptions.GoogleAPIError as exc:
-                # We failed to publish, set the exception on all futures and
-                # exit.
-                self._status = base.BatchStatus.ERROR
-
-                for future in self._futures:
-                    future.set_exception(exc)
-
-                _LOGGER.exception("Failed to publish %s messages.", len(self._futures))
-                return
+        # Begin the request to publish these messages.
+        # Log how long the underlying request takes.
+        start = time.time()
+
+        try:
+            response = self._client.api.publish(self._topic, self._messages)
+        except google.api_core.exceptions.GoogleAPIError as exc:
+            # We failed to publish, set the exception on all futures and
+            # exit.
+            self._status = base.BatchStatus.ERROR
```

Comment:
The `self._state_lock` is designed to protect setting `self._status`:  this change undoes that protection in the case that an error occurs, which introduces a race condition AFAICT.

Reply:
Hi Tres, we have covered that in the PR description:
> We only need to hold the state lock for the transition from ACCEPTING_MESSAGES / STARTING to IN_PROGRESS. After that, since only one thread is able to transition to IN_PROGRESS, we can safely release the state lock before calling the publish api and eventually transitioning to SUCCESS / ERROR.

Maybe you can elaborate on the race condition? Thanks

Reply:
TL; DR - Not holding the lock after changing the state to IN_PROGRESS indeed seems safe. However, I would still like to see the effect of this change covered by a test before merging. I can help with this, if needed.

---
**Long reply:**

I checked the Batch code, and from what I can tell, that the following threads are relevant:
- `MonitorBatchPublisher` (if autcommit == True): Triggers `_commit()` after `max_latency` time elapses.
- `CommitBatchPublisher`: If somebody calls `commit()` and the batch is in ACCEPTING_MESSAGES state, `_commit()` gets called in this thread.
- User thread(s): When calling `publisher.publish()`, `batch.publish()` gets called. If that causes an overflow, `commit()` is called in the `CommitBatchPublisher` thread.

The first thing that `_commit()` does after acquiring the lock is changing the batch status to IN_PROGRESS.

In that state, any further calls to `publish()` [become a no-op](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py#L280-L282), because the `will_accept()` check [returns `False`](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/publisher/_batch/base.py#L123-L124) for any state that is not ACCEPTING_MESSAGES.

Additionally, any other `_commit()` calls also [become a no-op](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py#L181-L188), because the batch status is now IN_PROGRESS.

Only the thread that changed the state to IN_PROGRESS can proceed, thus shielding the rest of the `_commit()` method with `_state_lock` indeed seems unnecessary, which includes the call to `self._client.api.publish()`.

_Without_ the `_state_lock` held, further calls to `batch.publish()` will not block longer than needed. The batch will simply not accept a new message, return `False`, and `publisher.publish() `will create an entirely [new batch](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/publisher/client.py#L236-L238) as a result.